src/main/java/org/breedinginsight/brapps/importer/model/base/Germplasm.java

```diff
@@ -103,11 +123,33 @@ public BrAPIGermplasm constructBrAPIGermplasm() {
                     .forEach(additionalInfo -> germplasm.putAdditionalInfoItem(additionalInfo.getAdditionalInfoName(), additionalInfo.getAdditionalInfoValue()));
         }
 
+        //TODO: add logic later for generating entry numbers if not provided by user
+        if (entryNo != null){
+            germplasm.putAdditionalInfoItem("ImportEntryNumber", entryNo);
+        }
+
+        //If there is an external uid, source is associated with it as an additional external reference
+        BrAPIExternalReference uidExternalReference = null;
+        if (germplasmSource != null) {
+            if (externalUID != null) {
+                uidExternalReference = new BrAPIExternalReference();
+                uidExternalReference.setReferenceID(getExternalUID());
+                uidExternalReference.setReferenceSource(getGermplasmSource());
+            } else {
+                germplasm.setSeedSourceDescription(getGermplasmSource());
+            }
+        }
+
         if (externalReferences != null) {
             List<BrAPIExternalReference> brAPIExternalReferences = externalReferences.stream()
                     .map(externalReference -> externalReference.constructBrAPIExternalReference())
                     .collect(Collectors.toList());
+            if (uidExternalReference != null) {
+                brAPIExternalReferences.add(uidExternalReference);
```

Comment:
I implemented it this way because the original code converts the external references stream to a list already (via collect(Collectors.ltoList(), so in the case where there isn't an external references mapping but we still want to add in the uid external reference we need to convert it to a list. I used singletonlist just cause it was a quick way to do that, but now I'm wondering if the restrictions of singletonList (immutable, one value) is the best choice here...

Reply:
Ah I see now. Yeah since its always going to be a list of one its probably not a big deal. I agree the immutable part might cause troubles later on if you want to add another external reference. I didn't test this out, just google it, but what about something like this?

```Arrays.asList(uidExternalReference)```

Reply:
That works, thanks!